### PR TITLE
Nette Framework

### PR DESCRIPTION
**Pull request by fprochazka**
Hello,
Nette Framework https://nette.org/ is quite popular in Czech Republic, that's also [why it got third in the PHP Frameworks 2015 survey on Sitepoint](http://www.sitepoint.com/best-php-framework-2015-sitepoint-survey-results/).

The framework includes an optional [Database](https://github.com/nette/database) component. Since It's optional and only a DBAL, not an ORM, I choose to use Doctrine like Symfony has in it's example. It's pretty common to use Doctrine with Nette.

## Remaining work
- [x] fix coding style
    - I've copied code from Symfony example, and from Nette sandbox, but both have different coding style
    - should I use PSR-1 & PSR-2, or the standard that Nette uses?
- [x] clean up dependencies
    - remove the repositories section from `composer.json`, after the required package is fixed
- [x] add tests

### CODE REVIEW

File: nette/notejam/app/Components/ForgottenPasswordControl.php [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+
+	public function formSucceeded(Form $form, $values)
+	{
+		if (!$user = $this->userRepository->findOneBy(['email' => $values->email])) {
+			$form['email']->addError("User with given email doesn't exist");
+			return;
+		}
+
+		// this is not a very secure way of getting new password
+		// but it's the samy way the symfony app is doing it...
```
Comment (jkuchar) [inline nette/notejam/app/Components/ForgottenPasswordControl.php]: // but it's the **same** way the symfony app is doing it...

File: nette/notejam/src/Notejam/Notes/Note.php [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+    /**
+     * @var \Datetime
+     *
+     * @ORM\Column(name="updated_at", type="datetime")
+     */
+    private $updatedAt;
+
+
+
+    public function __construct(User $user)
```
Comment (jkuchar) [inline nette/notejam/src/Notejam/Notes/Note.php]: Shouln't you require IIdentity here? (or check type in NoteControl:110)

File: nette/notejam/app/Components/NoteControl.php [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+		if ($values->pad === NULL) {
+			$note->setPad(NULL);
+
+		} else {
+			$note->setPad($this->padRepository->find($values->pad));
+		}
+
+		$this->em->flush();
+
+		$this->onSuccess($this);
```
Comment (jkuchar) [inline nette/notejam/app/Components/NoteControl.php]: explanation maybe?

// fires onSuccess event
// @see $this->onSuccess array

File: nette/notejam/src/Notejam/Notes/Note.php (hunk 2 above)
Comment (fprochazka) [inline nette/notejam/src/Notejam/Notes/Note.php]: Not really, since the identity is the entity. And the relation is for the entity, not for the relation.

**Comment by komarserjio** [comment]
Amazing. Just recently added Nette to a short list and now implementation is here!
I need some time to go through the implementation in details, but overall it looks very good. I see only minor things I'd change to keep consistency with other apps. Great job.

**Comment by klimesf** [comment]
By an accident, I started creating my Nette Notejam earlier today, probably around the same time as @fprochazka. We got challenged by a tweet ;)

Here is the result of my efforts (I'm not using Doctrine, but default Nette Database implementation).
As of now, I don't have any tests nor documentation.

https://github.com/klimesf/notejam

I will just leave it here. In case you are interested, feel free to contact me. If not, it was a good challenge for me at least.

**Comment by komarserjio** [comment]
Hey @klimesf.
It would be shameful to ignore your efforts. Of course we can include both implementations and specify in Nette's README that two alternative implementations are available.
Any chances to complete implementation with tests and docs?

**Comment by klimesf** [comment]
Hey @komarserjio, thanks for your kind reply!

Meanwhile, @fprochazka asked me to help him with the tests, if I wanted. I need some rest at the moment, but I will continue tomorrow. :)

**Comment by komarserjio** [comment]
Sure, no hurry :)
I hope you, guys, can share a lot to each other - things like templates, unit tests, route definitions, auth logic probably. Would be great if implementations will be as equal as possible. So people who tries Nette could easily compare Nette + Doctrine vs Nette + Native DB layer.

**Comment by klimesf** [comment]
Hey @komarserjio and @fprochazka,
I have added a few acceptance tests using Codeception. They cover the basic path through each use-case.

@fprochazka: Feel free to use any of these, if you like them :)

**Comment by klimesf** [comment]
Hey @komarserjio and @fprochazka,
I have extended acceptance tests to cover exceptional paths aswell and to match express example's ones. I have also added basic readme (inspired by @fprochazka 's one, hope you don't mind).

**Comment by komarserjio** [comment]
@klimesf Great, create a PR when you ready :)

**Comment by komarserjio** [comment]
Hey @fprochazka, great job is done!
I went through the app and have few comments.

Minor **functional bugs**:
* Impossible to create a note without a pad (Pad should not be required field)
* It is possible to create a pad with no name
* No confirmation page when deleting pad or note 
* No flash success message after pad or note is deleted
* No flash success message after password is changed
* No flash success message after pad is edited

**Coding style**
> should I use PSR-1 & PSR-2, or the standard that Nette uses?

I'm personally not absolutely happy about Nette's coding style, but it does not matter. I see Nette provides CS guide https://nette.org/en/coding-standard, so we have to accept and respect this. So please make sure the codebase follows Nette's coding standard. Also I would add docblock section to each method (of course in a format which is recommended by CS guide).

**Javascript**
What netteForms.min.js file is used for? If it is not absolutely necessary I would avoid using any additional javascript code.

**Tests**
I would reuse tests provided by @klimesf.

**Comment by fprochazka** [comment]
@komarserjio thanks for the feedback, here is some progress and responses.

## Bugs

> Impossible to create a note without a pad (Pad should not be required field)

fixed

> It is possible to create a pad with no name

fixed

> No confirmation page when deleting pad or note

There is also none in the symfony implementation, which I was using as a template for the functionality.
Also, is it really needed? If so, why? Is it only a protection for CSRF?
Nette doesn't need to use a form for that. I'd preffer simply using JS confirmation dialog.
The protection is handled using the `nextras/secured-links` package.

> flash messages

I've added them and some more

## Coding style

> I'm personally not absolutely happy about Nette's coding style, but it does not matter.

It's basicaly PSR1 & PSR2, but with tabs. I've reformated it & added docblocks.

## Javascript

This is a part of nette, that is provided by nette/forms. I just didn't wanna add another package to solve the propagation of the file from package to document root (Nette's variant of Assetic). It handles browser-side validations. Nette forms generate rules to data attributes from the validations in your code and the js then enforces them in browser. So I'd argue it's pretty vital.

**Comment by komarserjio** [comment]
@fprochazka, thanks for the update!

Regarding the confirmation page. 
Yes, it's needed. It's not about CSRF. The main motivation is to show how a framework handles requests based on types. So the delete url is `/notes/1/delete` and depends on type (`GET` or `POST`) different logic should be applied. In case of `GET` confirmation page is shown, in case of `POST` a note should be deleted. Some frameworks solve this on routing level, some frameworks does not care about this and you need manually detect type in an action and react accordingly, like:
```
if (type == POST) {
    ... delete a note and redirect
} else {
    ... show confirmation page
}
```
So curious how to act with Nette in this case :).

Probably something is broken with Symfony app after upgrade to 2.7. Sorry about that. Take a look on cakephp or laravel version to see the confirmation page.

**Comment by fprochazka** [comment]
@komarserjio okey, that makes sense :) I will add the confirmation pages.

**Comment by fprochazka** [comment]
I've added the confirmation pages, now only tests remain.

**Comment by fprochazka** [comment]
@komarserjio Okey, I've copied the tests from @klimesf (thank you!) and fixed them (and the app). So now it should be almost done, after I cleanup the dependencies.

**Comment by klimesf** [comment]
@fprochazka you are most certainly welcome ;) Were there some serious fixes in the tests I should know about? I can't see changes in ba72a6d.

**Comment by komarserjio** [comment]
Hey guys @fprochazka @klimesf.
Both implementations are very close to be merged. It would be great if you could carefully review each other's implementation. Since I have no production with the Nette I'm bad reviewer, I can argue only on general PHP things.

**Comment by klimesf** [comment]
I believe @fprochazka is above my level of experience with Nette ;) however, I will surely go through his implementation, just for the sake of inspiration.

@fprochazka: I would love have my code reviewed by you! I know you don't have much time, so I leave it up to you.

**Comment by fprochazka** [comment]
> Were there some serious fixes in the tests I should know about?

I had slightly different wording on few places and one different column in db (I copied entities from symfony). So it wasn't fixing really, more like adjusting to work with my impl :)

**Comment by fprochazka** [comment]
@komarserjio @klimesf Okey, I've fixed last thing that bugged me. This should be ready for review and then merging :)

**Comment by fprochazka** [comment]
@komarserjio what is the procedure now? :)

**Comment by komarserjio** [comment]
Next step is merge :)
@klimesf do you have any comments after the review of this implementation?

**Comment by klimesf** [comment]
@komarserjio I got nothing, it's perfect :)

**Comment by komarserjio** [comment]
Both PRs use `/nette` folder :) We need to solve this somehow.
I'd suggest @fprochazka's implementation goes under `/nette/doctrine/` and @klimesf under `/nette/native_db`. 
On top `/nette` level I will create README file with explanation that two implementation are available.
Thougths?

**Comment by klimesf** [comment]
@komarserjio: I have thought about this a while ago. I could put my implementation under `nette-native-db` directory, @fprochazka under `nette-doctrine` and mention it in the root README, or we could go the way you suggested. Since this is your repository, it's your call. ;)

**Comment by fprochazka** [comment]
@komarserjio I don't mind either way :)

**Comment by komarserjio** [comment]
Ok, @fprochazka could you please move the app to `/nette/doctrine` (and update the path in your readme).

**Comment by fprochazka** [comment]
@komarserjio like this? :)

**Comment by fprochazka** [comment]
@komarserjio thank you! 

snimek obrazovky porizeny 2015-12-09 11 58 07

This is intentional? We might wanna add another readme to the repository in nette dir as crossroads, so people notice there are two impl. Should we help with that? :)

**Comment by komarserjio** [comment]
There is an issue with markdown. I will fix that :)

**Comment by fprochazka** [comment]
@komarserjio awesome, thank you :)

**Pull request merged**